### PR TITLE
Add the syntax for abstract indexes.

### PR DESCRIPTION
**Pull request by vpetrovykh**
Add the syntax for complex abstract index definitions as well as update the concrete index syntax to account for more index variety.

### CODE REVIEW

**Comment by vpetrovykh** [comment]
I've also updated the related [RFC 1017](https://github.com/edgedb/rfcs/pull/66).

File: edb/edgeql/ast.py [hunk 1]
```diff
@@ -1170,24 +1170,54 @@ class DropConcreteConstraint(ConcreteConstraintOp, DropObject):
     pass
 
 
+class IndexType(DDL):
+    name: typing.Union[tuple, str]
```
Comment (msullivan) [inline edb/edgeql/ast.py:1174]: It's a `tuple[str]`, right?

File: edb/edgeql/codegen.py [hunk 2]
```diff
@@ -1808,19 +1817,82 @@ def _after_index(self, node: qlast.IndexCommand) -> None:
...
+            for i, (name, arg) in enumerate(node.kwargs.items()):
+                if i > 0:
+                    self.write(', ')
+                self.write(f'{edgeql_quote.quote_ident(name)} := ')
+                self.visit(arg)
+            self.write(')')
+
     def visit_CreateIndex(self, node: qlast.CreateIndex) -> None:
+        def after_name() -> None:
+            if node.params:
```
Comment (msullivan) [inline edb/edgeql/codegen.py:1841]: What are params, here? How do they interact with kwargs?

File: edb/edgeql/parser/grammar/commondl.py [hunk 3]
```diff
@@ -543,3 +561,111 @@ def reduce_ExtensionVersion(self, *kids):
...
+
+    def reduce_IndexExtArgList(self, *kids):
+        self.val = kids[0].val
+
+    def reduce_empty(self):
+        self.val = []
+
+
+class ProcessIndexArgsMixin:
+    def _process_arguments(self, arguments):
```
Comment (msullivan) [inline edb/edgeql/parser/grammar/commondl.py:622]: This could just be a function, I think, instead of a mixin?

Comment (msullivan) [reply edb/edgeql/parser/grammar/commondl.py:622]: Oh but there's an existing one for functions?

File: edb/edgeql/codegen.py (hunk 2 above)
Comment (msullivan) [reply edb/edgeql/codegen.py:1841]: Oh, I understand now. Only one can be present, though, right? Maybe have an assert to that effect?

**Comment by stealth-sparrow** [comment]
Is it possible to add support for `array_ops` indexes?

```
CREATE ABSTRACT INDEX std::gin_array ON array<anytype> {
    USING SQL $$gin ((__col__))$$;
};
```

**Comment by stealth-toucan** [comment]
Is it possible to add support for `array_ops` indexes?

```
CREATE ABSTRACT INDEX std::gin_array ON array<anytype> {
    USING SQL $$gin ((__col__))$$;
};
```

File: edb/edgeql/parser/grammar/commondl.py (hunk 3 above)
Comment (vpetrovykh) [reply edb/edgeql/parser/grammar/commondl.py:622]: At the moment I'm just following the same pattern as for functions.

File: edb/edgeql/codegen.py (hunk 2 above)
Comment (vpetrovykh) [reply edb/edgeql/codegen.py:1845]: No args or params in this iteration of the PR anymore.

**Comment by vpetrovykh** [comment]
This implementation hard-codes the valid indexable types as well as the specific SQL that needs to be emitted to define the built-in Postgres indexes. In the future both of these may be defined more formally via DDL, though.

**Review by msullivan — APPROVED** [review]
This looks good to me. Can you add some DDL tests?

**Review by msullivan — APPROVED** [review]

**Pull request merged**